sdk/identity/Azure.Identity/src/Credentials/DefaultAzureCredentialOptions.cs

```diff
@@ -11,12 +14,90 @@ namespace Azure.Identity
     /// </summary>
     public class DefaultAzureCredentialOptions : TokenCredentialOptions
     {
+        private struct UpdateTracker<T>
+        {
+            private bool _updated;
+            private T _value;
+
+            public UpdateTracker(T initialValue)
+            {
+                _value = initialValue;
+                _updated = false;
+            }
+
+            public T Value
+            {
+                get => _value;
+                set
+                {
+                    _value = value;
+                    _updated = true;
+                }
+            }
+
+            public bool Updated => _updated;
+        }
+
+        private UpdateTracker<string> _tenantId = new UpdateTracker<string>(GetNonEmptyStringOrNull(EnvironmentVariables.TenantId));
+        private UpdateTracker<string> _interactiveBrowserTenantId = new UpdateTracker<string>(GetNonEmptyStringOrNull(EnvironmentVariables.TenantId));
+        private UpdateTracker<string> _sharedTokenCacheTenantId = new UpdateTracker<string>(GetNonEmptyStringOrNull(EnvironmentVariables.TenantId));
+        private UpdateTracker<string> _visualStudioTenantId = new UpdateTracker<string>(GetNonEmptyStringOrNull(EnvironmentVariables.TenantId));
+        private UpdateTracker<string> _visualStudioCodeTenantId = new UpdateTracker<string>(GetNonEmptyStringOrNull(EnvironmentVariables.TenantId));
+
+        /// <summary>
+        /// The ID of the tenant to which the credential will authenticate by default. If not specified, the credential will authenticate to any requested tenant, and will default to the tenant to which the chosen authentication method was originally authenticated.
+        /// </summary>
+        public string TenantId
+        {
+            get => _tenantId.Value;
+            set
+            {
+                if (_interactiveBrowserTenantId.Updated && value != _interactiveBrowserTenantId.Value)
+                {
+                    throw new InvalidOperationException("Applications should not set both TenantId and InteractiveBrowserTenantId. TenantId is preferred, and is functionally equivalent. InteractiveBrowserTenantId exists only to provide backwards compatibility.");
```

Comment:
possible simplification: It might be nice to implement a method on `UpdateTracker` like `ThrowIfChangedFrom(UpdateTracker other)` and centralize all this throw logic. You'd just need an extra property to store the friendly name of the property so that this exception message could be interpolated.

Reply:
We could possibly consolidate this logic, but my first crack at this was a bit hacky. I'm tempted to leave the straight forward implementation here and if you feel it should be simplified create a follow up work-item. Thoughts?

Reply:
I think leaving it as-is is fine.